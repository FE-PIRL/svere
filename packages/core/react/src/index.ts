import React, { useRef, useEffect, useState } from "react";
import resolveWrapperProps from "../../utils";

export default (Component: any, wrapperProps?: WrapperProps) => (props: {
  [x: string]: any;
}) => {
  const container = useRef<HTMLElement>(null);
  const component = useRef<HTMLElement | null>(null);

  const [mounted, setMount] = useState(false);

  useEffect(() => {
    const onRegex = /on([A-Z]{1,}[a-zA-Z]*)/;
    const watchRegex = /watch([A-Z]{1,}[a-zA-Z]*)/;

    component.current = new Component({ target: container.current, props });

    let watchers: any[][] = [];
    for (const key in props) {
      const onMatch = key.match(onRegex);
      const watchMatch = key.match(watchRegex);

      if (onMatch && typeof props[key] === "function") {
        ((component.current as unknown) as SvelteComponent)?.$on(
          `${onMatch[1][0].toLowerCase()}${onMatch[1].slice(1)}`,
          props[key]
        );
      }

      if (watchMatch && typeof props[key] === "function") {
        watchers.push([
          `${watchMatch[1][0].toLowerCase()}${watchMatch[1].slice(1)}`,
          props[key],
        ]);
      }
    }

    if (watchers.length) {
      const update = ((component.current as unknown) as SvelteComponent)?.$$
        .update;
      if (update) {
        ((component.current as unknown) as SvelteComponent).$$.update = function () {
          watchers.forEach(([name, callback]) => {
            const index = ((component.current as unknown) as SvelteComponent)
              ?.$$.props[name];
            const prop = ((component.current as unknown) as SvelteComponent)?.$$
              .ctx[index];
            prop && callback(prop);
          });
          update.apply(null, arguments);
        };
      }
    }

    return () => {
      ((component.current as unknown) as SvelteComponent)?.$destroy();
    };
  }, []);

  useEffect(() => {
    if (!mounted) {
      setMount(true);
      return;
    }

    ((component.current as unknown) as SvelteComponent)?.$set(props);
  }, [props]);

  wrapperProps = resolveWrapperProps(wrapperProps);

  return React.createElement(wrapperProps?.element, {
    ref: container,
    id: wrapperProps?.id,
    className: wrapperProps?.className,
    style: { ...wrapperProps?.styles },
  });
};
