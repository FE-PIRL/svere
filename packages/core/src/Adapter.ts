import React, { useRef, useEffect, useState } from "react";
import Vue from "vue";

const defaultWrapperProps = {
  element: "div",
  id: "svelte-wrapper",
};

export const toReact = (
  Component: any,
  wrapperProps?: WrapperProps
) => (props: { [x: string]: any }) => {
  const container = useRef<HTMLElement>(null);
  const component = useRef<HTMLElement | null>(null);

  const [mounted, setMount] = useState(false);

  useEffect(() => {
    const onRegex = /on([A-Z]{1,}[a-zA-Z]*)/;
    const watchRegex = /watch([A-Z]{1,}[a-zA-Z]*)/;

    component.current = new Component({ target: container.current, props });

    let watchers: any[][] = [];
    for (const key in props) {
      const onMatch = key.match(onRegex);
      const watchMatch = key.match(watchRegex);

      if (onMatch && typeof props[key] === "function") {
        ((component.current as unknown) as SvelteComponent)?.$on(
          `${onMatch[1][0].toLowerCase()}${onMatch[1].slice(1)}`,
          props[key]
        );
      }

      if (watchMatch && typeof props[key] === "function") {
        watchers.push([
          `${watchMatch[1][0].toLowerCase()}${watchMatch[1].slice(1)}`,
          props[key],
        ]);
      }
    }

    if (watchers.length) {
      const update = ((component.current as unknown) as SvelteComponent)?.$$
        .update;
      if (update) {
        ((component.current as unknown) as SvelteComponent).$$.update = function () {
          watchers.forEach(([name, callback]) => {
            const index = ((component.current as unknown) as SvelteComponent)
              ?.$$.props[name];
            const prop = ((component.current as unknown) as SvelteComponent)?.$$
              .ctx[index];
            prop && callback(prop);
          });
          update.apply(null, arguments);
        };
      }
    }

    return () => {
      ((component.current as unknown) as SvelteComponent)?.$destroy();
    };
  }, []);

  useEffect(() => {
    if (!mounted) {
      setMount(true);
      return;
    }

    ((component.current as unknown) as SvelteComponent)?.$set(props);
  }, [props]);

  if (!wrapperProps) {
    wrapperProps = defaultWrapperProps;
  } else {
    wrapperProps = Object.assign({}, defaultWrapperProps, wrapperProps);
  }

  return React.createElement(wrapperProps.element, {
    ref: container,
    id: wrapperProps.id,
    className: wrapperProps.className,
    style: { ...wrapperProps.styles },
  });
};

export const toVue = (Component: any, wrapperProps?: WrapperProps) => (props: {
  [x: string]: any;
}) => {
  if (!wrapperProps) {
    wrapperProps = defaultWrapperProps;
  } else {
    wrapperProps = Object.assign({}, defaultWrapperProps, wrapperProps);
  }

  return Vue.component(wrapperProps.id as string, {
    render(createElement) {
      return createElement(wrapperProps?.element, {
        ref: "container",
        attrs: { class: wrapperProps?.className, id: wrapperProps?.id },
        style: { ...wrapperProps?.styles },
      });
    },
    data() {
      return {
        comp: null,
      };
    },
    mounted() {
      (this as any).comp = new Component({
        target: this.$refs.container,
        props,
      });

      let watchers: (string | Function | Function[])[][] = [];

      for (const key in this.$listeners) {
        (this as any).comp.$on(key, this.$listeners[key]);
        const watchRe = /watch:([^]+)/;

        const watchMatch = key.match(watchRe);

        if (watchMatch && typeof this.$listeners[key] === "function") {
          watchers.push([
            `${watchMatch[1][0].toLowerCase()}${watchMatch[1].slice(1)}`,
            this.$listeners[key],
          ]);
        }
      }

      if (watchers.length) {
        let comp = (this as any).comp;
        const update = (this as any).comp.$$.update;

        (this as any).comp.$$.update = function () {
          watchers.forEach(([name, callback]) => {
            // @ts-ignore
            const index = comp.$$.props[name];
            const prop = comp.$$.ctx[index];
            // @ts-ignore
            prop && callback(prop);
          });
          update.apply(null, arguments);
        };
      }
    },
    updated() {
      (this as any).comp.$set(this.$attrs);
    },
    destroyed() {
      (this as any).comp.$destroy();
    },
  });
};
