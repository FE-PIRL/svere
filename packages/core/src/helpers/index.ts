const resolveWrapperProps = (wrapperProps?: WrapperProps) => ({
    element: "div",
    id: "svelte-wrapper",
    ...wrapperProps
});
export default resolveWrapperProps