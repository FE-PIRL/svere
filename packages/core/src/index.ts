export { toReact, toVue } from "./Adapter";
