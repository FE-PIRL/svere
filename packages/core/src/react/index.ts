import React, { useRef, useEffect, useState }  from "react";
import resolveWrapperProps from "../helpers";

export default (Component: SvelteComponentConstructor, _wrapperProps?: WrapperProps) => (props: {
  [x: string]: any;
}) => {
  const container = useRef<HTMLElement>(null);
  const component = useRef<SvelteComponent | null>(null);

  const [mounted, setMount] = useState(false);

  useEffect(() => {
    const onRegex = /on([A-Z]+[a-zA-Z]*)/;
    const watchRegex = /watch([A-Z]+[a-zA-Z]*)/;

    component.current = new Component({ target: container.current, props });

    let watchers: any[][] = [];
    for (const key in props) {
      const onMatch = key.match(onRegex);
      const watchMatch = key.match(watchRegex);

      if (onMatch && typeof props[key] === "function") {
        ((component.current as unknown) as SvelteComponent)?.$on(
            `${onMatch[1][0].toLowerCase()}${onMatch[1].slice(1)}`,
            props[key]
        );
      }

      if (watchMatch && typeof props[key] === "function") {
        watchers.push([
          `${watchMatch[1][0].toLowerCase()}${watchMatch[1].slice(1)}`,
          props[key],
        ]);
      }
    }

    if (watchers.length) {
      if (component.current?.$$.update) {
        const update = component.current?.$$.update;
        component.current.$$.update = function () {
          watchers.forEach(([name, callback]) => {
            const index = component.current?.$$.props[name];
            const prop = component.current?.$$.ctx[index];
            prop && callback(prop);
          });
          update.apply(null, arguments);
        };
      }
    }

    return () => component.current?.$destroy();
  }, []);

  useEffect(() => {
    if (!mounted) {
      setMount(true);
      return;
    }

    component.current?.$set(props);
  }, [props]);

  const wrapperProps = resolveWrapperProps(_wrapperProps);

  return React.createElement(wrapperProps.element, {
    ref: container,
    ...wrapperProps
  });
};
