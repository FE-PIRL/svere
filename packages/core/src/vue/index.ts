import Vue from "vue";
import resolveWrapperProps from "../helpers";

export default (Component: any, wrapperProps?: WrapperProps) => {
    wrapperProps = resolveWrapperProps(wrapperProps);

    return Vue.component(wrapperProps.id as string, {
        render(createElement) {
            return createElement(wrapperProps?.element, {
                ref: "container",
                class: wrapperProps?.className,
                attrs: { id: wrapperProps?.id },
                style: { ...wrapperProps?.styles },
            });
        },
        data() {
            return {
                comp: null,
            };
        },
        mounted() {
            this.comp = new Component({
                target: this.$refs.container,
                props: this.$attrs,
            });

            let watchers: any[][] = [];

            for (const key in this.$listeners) {
                (this as any).comp.$on(key, this.$listeners[key]);

                const watchRe = /watch:([^]+)/;
                const watchMatch = key.match(watchRe);

                if (watchMatch && typeof this.$listeners[key] === "function") {
                    watchers.push([
                        `${watchMatch[1][0].toLowerCase()}${watchMatch[1].slice(1)}`,
                        this.$listeners[key],
                    ]);
                }
            }

            if (watchers.length) {
                let comp = (this as any).comp;
                const update = (this as any).comp.$$.update;

                (this as any).comp.$$.update = function () {
                    watchers.forEach(([name, callback]) => {
                        const index = comp.$$.props[name];
                        const prop = comp.$$.ctx[index];
                        prop && callback(prop);
                    });
                    update.apply(null, arguments);
                };
            }
        },
        updated() {
            (this as any).comp.$set(this.$attrs);
        },
        destroyed() {
            (this as any).comp.$destroy();
        },
    });
};
