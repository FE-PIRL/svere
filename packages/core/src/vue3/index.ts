import {defineComponent, h, onDeactivated, onMounted, onUpdated, ref} from "vue";
import resolveWrapperProps from "../helpers";

export default (Component: SvelteComponentConstructor, _wrapperProps?: WrapperProps) => {
    const wrapperProps = resolveWrapperProps(_wrapperProps);

    return defineComponent({
        name: wrapperProps.id,
        setup(props, ctx) {
            const container = ref<HTMLElement | null>(null);
            const component = ref<SvelteComponent | null>(null);

            onMounted(() => {
                const onRegex = /on([A-Z]+[a-zA-Z]*)/;
                const watchRegex = /watch([A-Z]+[a-zA-Z]*)/;

                component.value = new Component({
                    target: container.value,
                    props: ctx.attrs,
                });

                let watchers: any[][] = [];
                for (const key in ctx.attrs) {
                    const onMatch = key.match(onRegex);
                    const watchMatch = key.match(watchRegex);

                    if (onMatch && typeof ctx.attrs[key] === "function") {
                        component.value.$on(
                            `${onMatch[1][0].toLowerCase()}${onMatch[1].slice(1)}`,
                            props[key]
                        );
                    }

                    if (watchMatch && typeof props[key] === "function") {
                        watchers.push([
                            `${watchMatch[1][0].toLowerCase()}${watchMatch[1].slice(1)}`,
                            props[key],
                        ]);
                    }
                }

                if (watchers.length) {
                    if (component.value.$$.update) {
                        const update = component.value.$$.update;
                        component.value.$$.update = function () {
                            watchers.forEach(([name, callback]) => {
                                const index = component.value.$$.props[name];
                                const prop = component.value.$$.ctx[index];
                                prop && callback(prop);
                            });
                            update.apply(null, arguments);
                        };
                    }
                }

                return () => component.value.$destroy();
            });

            onUpdated(() => component.value?.$set(ctx.attrs));

            onDeactivated(() => component.value?.$destroy());

            return {
                component,
                container,
            };
        },
        render() {
            const children = {
                ref: "container",
                id: wrapperProps.id,
                style: {...wrapperProps.styles},
                class: wrapperProps.className,
            };
            return h(wrapperProps.element, children);
        },
    });
};
