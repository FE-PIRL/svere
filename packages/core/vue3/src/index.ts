import { h, defineComponent, ref, onMounted, onUpdated, onDeactivated } from "vue";
import resolveWrapperProps from "../../utils";

export default (Component: any, wrapperProps?: WrapperProps) => {
    wrapperProps = resolveWrapperProps(wrapperProps);

    return defineComponent({
        name: wrapperProps.id as string,
        setup(props, ctx) {
            const container = ref(null);
            const component = ref(null);

            onMounted(() => {
                const onRegex = /on([A-Z]{1,}[a-zA-Z]*)/;
                const watchRegex = /watch([A-Z]{1,}[a-zA-Z]*)/;

                component.value = new Component({
                    target: container.value,
                    props: ctx.attrs,
                });

                let watchers: any[][] = [];
                for (const key in ctx.attrs) {
                    const onMatch = key.match(onRegex);
                    const watchMatch = key.match(watchRegex);

                    if (onMatch && typeof ctx.attrs[key] === "function") {
                        ((component.value as unknown) as SvelteComponent)?.$on(
                            `${onMatch[1][0].toLowerCase()}${onMatch[1].slice(1)}`,
                            props[key]
                        );
                    }

                    if (watchMatch && typeof props[key] === "function") {
                        watchers.push([
                            `${watchMatch[1][0].toLowerCase()}${watchMatch[1].slice(1)}`,
                            props[key],
                        ]);
                    }
                }

                if (watchers.length) {
                    const update = ((component.value as unknown) as SvelteComponent)?.$$
                        .update;
                    if (update) {
                        ((component.value as unknown) as SvelteComponent).$$.update = function () {
                            watchers.forEach(([name, callback]) => {
                                const index = ((component.value as unknown) as SvelteComponent)
                                    ?.$$.props[name];
                                const prop = ((component.value as unknown) as SvelteComponent)
                                    ?.$$.ctx[index];
                                prop && callback(prop);
                            });
                            update.apply(null, arguments);
                        };
                    }
                }

                return () => {
                    ((component.value as unknown) as SvelteComponent)?.$destroy();
                };
            });

            onUpdated(() => {
                ((component.value as unknown) as SvelteComponent)?.$set(ctx.attrs);
            });

            onDeactivated(()=>{
                ((component.value as unknown) as SvelteComponent)?.$destroy();
            });

            return {
                component,
                container,
            };
        },
        render() {
            const childrenDefaults = {
                ref: "container",
                id: wrapperProps?.id,
                style: { ...wrapperProps?.styles },
            };
            const children = wrapperProps?.className
                ? Object.assign({}, childrenDefaults, {
                    class: wrapperProps?.className,
                })
                : childrenDefaults;
            return h(wrapperProps?.element, children);
        },
    });
};
