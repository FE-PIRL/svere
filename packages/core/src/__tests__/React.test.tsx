import React from "react";
import ReactDOM from "react-dom";

import { reactify } from "../../src/index";
import SvelteComponent from "../../../../examples/dist";
import { act } from "react-dom/test-utils";

describe("React", () => {
  let container: any;
  const name = "ben";

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  it("should render", async () => {
    const ReactComponent = reactify(SvelteComponent);

    // render the component
    act(() => {
      ReactDOM.render(<ReactComponent name={name} />, container);
    });

    expect(container).toMatchSnapshot();
  });

  it("count increments when button is clicked", async () => {
    const ReactComponent = reactify(SvelteComponent);

    // render the component
    act(() => {
      ReactDOM.render(<ReactComponent name={name} />, container);
    });

    const button = container.querySelector("button");
    expect(button.textContent).toBe("count: 0");

    // Test second render and componentDidUpdate
    await act(async () => {
      button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });

    expect(button.textContent).toBe("count: 1");
  });

  it("listen custom event", async () => {
    const ReactComponent = reactify(SvelteComponent);

    const handleSomeEvent = (e: any) => {
      expect(e.detail).toBe(1);
    };

    // render the component
    await act(async () => {
      ReactDOM.render(
        <ReactComponent name={name} onSomeEvent={handleSomeEvent} />,
        container
      );
    });

    const button = container.querySelector("button");

    // Test second render and componentDidUpdate
    await act(async () => {
      button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
  });
});
