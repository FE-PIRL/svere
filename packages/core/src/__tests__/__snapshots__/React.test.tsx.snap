// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React reactify 1`] = `
<div>
  <span>
    <main
      class="svelte-1tky8bj"
    >
      <h1
        class="svelte-1tky8bj"
      >
        Hello 
        ben
        !
      </h1>
       
      <p>
        Visit the 
        <a
          href="https://svelte.dev/tutorial"
        >
          Svelte tutorial
        </a>
         to learn how to build Svelte apps.
      </p>
    </main>
  </span>
</div>
`;
